File: python/src/g3p_vm_gpu/builtins.py
from __future__ import annotations

from typing import List, Union

from .errors import Err, ErrCode
from .ast import Val


def _is_num(v: Val) -> bool:
    # bool is NOT numeric in this subset
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def _promote(a: Union[int, float], b: Union[int, float]) -> tuple[Union[int, float], Union[int, float], type]:
    if isinstance(a, float) or isinstance(b, float):
        return float(a), float(b), float
    return int(a), int(b), int


def builtin_call(name: str, args: List[Val]) -> Val | Err:
    if name == "abs":
        if len(args) != 1:
            return Err(ErrCode.TYPE, "abs expects 1 argument")
        x = args[0]
        if not _is_num(x):
            return Err(ErrCode.TYPE, "abs expects a numeric argument")
        return -x if x < 0 else x

    if name in ("min", "max"):
        if len(args) != 2:
            return Err(ErrCode.TYPE, f"{name} expects 2 arguments")
        a, b = args
        if not _is_num(a) or not _is_num(b):
            return Err(ErrCode.TYPE, f"{name} expects numeric arguments")
        a2, b2, _t = _promote(a, b)
        return a2 if (a2 <= b2 if name == "min" else a2 >= b2) else b2

    if name == "clip":
        if len(args) != 3:
            return Err(ErrCode.TYPE, "clip expects 3 arguments: clip(x, lo, hi)")
        x, lo, hi = args
        if not _is_num(x) or not _is_num(lo) or not _is_num(hi):
            return Err(ErrCode.TYPE, "clip expects numeric arguments")
        # promote all to float if any float appears
        any_float = any(isinstance(v, float) for v in (x, lo, hi))
        if any_float:
            x, lo, hi = float(x), float(lo), float(hi)
        else:
            x, lo, hi = int(x), int(lo), int(hi)
        if lo > hi:
            return Err(ErrCode.VALUE, "clip requires lo <= hi")
        if x < lo:
            return lo
        if x > hi:
            return hi
        return x

    if name == "len":
        if len(args) != 1:
            return Err(ErrCode.TYPE, "len expects 1 argument")
        x = args[0]
        if not isinstance(x, (str, list)):
            return Err(ErrCode.TYPE, "len expects string/list argument")
        return len(x)

    if name == "concat":
        if len(args) != 2:
            return Err(ErrCode.TYPE, "concat expects 2 arguments")
        a, b = args
        if isinstance(a, str) and isinstance(b, str):
            return a + b
        if isinstance(a, list) and isinstance(b, list):
            return a + b
        return Err(ErrCode.TYPE, "concat expects (string,string) or (list,list)")

    if name == "slice":
        if len(args) != 3:
            return Err(ErrCode.TYPE, "slice expects 3 arguments: slice(x, lo, hi)")
        x, lo, hi = args
        if not isinstance(x, (str, list)):
            return Err(ErrCode.TYPE, "slice expects string/list as first argument")
        if not (isinstance(lo, int) and not isinstance(lo, bool) and isinstance(hi, int) and not isinstance(hi, bool)):
            return Err(ErrCode.TYPE, "slice expects integer lo/hi")
        return x[lo:hi]

    if name == "index":
        if len(args) != 2:
            return Err(ErrCode.TYPE, "index expects 2 arguments: index(x, i)")
        x, i = args
        if not isinstance(x, (str, list)):
            return Err(ErrCode.TYPE, "index expects string/list as first argument")
        if not (isinstance(i, int) and not isinstance(i, bool)):
            return Err(ErrCode.TYPE, "index expects integer index")
        n = len(x)
        j = i + n if i < 0 else i
        if j < 0 or j >= n:
            return Err(ErrCode.VALUE, "index out of range")
        return x[j]

    return Err(ErrCode.NAME, f"unknown builtin: {name}")

File: cpp/src/builtins.cpp
#include "g3pvm/builtins.hpp"

namespace g3pvm {

namespace {

bool to_numeric_pair(const Value& a, const Value& b, double& a_out, double& b_out,
                     bool& any_float) {
  if (!is_numeric(a) || !is_numeric(b)) {
    return false;
  }
  any_float = (a.tag == ValueTag::Float) || (b.tag == ValueTag::Float);
  a_out = (a.tag == ValueTag::Float) ? a.f : static_cast<double>(a.i);
  b_out = (b.tag == ValueTag::Float) ? b.f : static_cast<double>(b.i);
  return true;
}

BuiltinResult fail(ErrCode code, const std::string& message) {
  BuiltinResult out;
  out.is_error = true;
  out.err = Err{code, message};
  return out;
}

}  // namespace

BuiltinResult builtin_call(const std::string& name, const std::vector<Value>& args) {
  if (name == "abs") {
    if (args.size() != 1) {
      return fail(ErrCode::Type, "abs expects 1 argument");
    }
    const Value& x = args[0];
    if (!is_numeric(x)) {
      return fail(ErrCode::Type, "abs expects a numeric argument");
    }
    BuiltinResult out;
    out.value = (x.tag == ValueTag::Float) ? Value::from_float(x.f < 0 ? -x.f : x.f)
                                            : Value::from_int(x.i < 0 ? -x.i : x.i);
    return out;
  }

  if (name == "min" || name == "max") {
    if (args.size() != 2) {
      return fail(ErrCode::Type, name + " expects 2 arguments");
    }
    double a = 0.0;
    double b = 0.0;
    bool any_float = false;
    if (!to_numeric_pair(args[0], args[1], a, b, any_float)) {
      return fail(ErrCode::Type, name + " expects numeric arguments");
    }
    const double pick = (name == "min") ? (a <= b ? a : b) : (a >= b ? a : b);
    BuiltinResult out;
    out.value = any_float ? Value::from_float(pick)
                          : Value::from_int(static_cast<long long>(pick));
    return out;
  }

  if (name == "clip") {
    if (args.size() != 3) {
      return fail(ErrCode::Type, "clip expects 3 arguments: clip(x, lo, hi)");
    }
    const Value& x = args[0];
    const Value& lo = args[1];
    const Value& hi = args[2];
    if (!is_numeric(x) || !is_numeric(lo) || !is_numeric(hi)) {
      return fail(ErrCode::Type, "clip expects numeric arguments");
    }
    const bool any_float =
        (x.tag == ValueTag::Float) || (lo.tag == ValueTag::Float) || (hi.tag == ValueTag::Float);
    if (any_float) {
      const double x2 = (x.tag == ValueTag::Float) ? x.f : static_cast<double>(x.i);
      const double lo2 = (lo.tag == ValueTag::Float) ? lo.f : static_cast<double>(lo.i);
      const double hi2 = (hi.tag == ValueTag::Float) ? hi.f : static_cast<double>(hi.i);
      if (lo2 > hi2) {
        return fail(ErrCode::Value, "clip requires lo <= hi");
      }
      BuiltinResult out;
      if (x2 < lo2) {
        out.value = Value::from_float(lo2);
      } else if (x2 > hi2) {
        out.value = Value::from_float(hi2);
      } else {
        out.value = Value::from_float(x2);
      }
      return out;
    }

    const long long x2 = x.i;
    const long long lo2 = lo.i;
    const long long hi2 = hi.i;
    if (lo2 > hi2) {
      return fail(ErrCode::Value, "clip requires lo <= hi");
    }
    BuiltinResult out;
    if (x2 < lo2) {
      out.value = Value::from_int(lo2);
    } else if (x2 > hi2) {
      out.value = Value::from_int(hi2);
    } else {
      out.value = Value::from_int(x2);
    }
    return out;
  }

  return fail(ErrCode::Name, "unknown builtin: " + name);
}

}  // namespace g3pvm
